src/runtime/pythonengine.cs

```diff
@@ -411,33 +411,54 @@ public static PyObject ModuleFromString(string name, string code)
         /// executing the code string as a PyObject instance, or null if
         /// an exception was raised.
         /// </remarks>
-        public static PyObject RunString(string code)
+        public static PyObject RunString(
+            string code, IntPtr? globals = null, IntPtr? locals = null
+            )
         {
-            IntPtr globals = Runtime.PyEval_GetGlobals();
-            IntPtr locals = Runtime.PyDict_New();
-
-            IntPtr builtins = Runtime.PyEval_GetBuiltins();
-            Runtime.PyDict_SetItemString(locals, "__builtins__", builtins);
+            bool borrowedGlobals = true;
+            if (globals == null)
+            {
+                globals = Runtime.PyEval_GetGlobals();
+                if (globals == IntPtr.Zero)
+                {
+                    globals = Runtime.PyDict_New();
+                    Runtime.PyDict_SetItemString(
+                        globals.Value, "__builtins__",
+                        Runtime.PyEval_GetBuiltins()
+                        );
+                    borrowedGlobals = false;
+                }
+            }
 
-            IntPtr flag = (IntPtr)257; /* Py_file_input */
-            IntPtr result = Runtime.PyRun_String(code, flag, globals, locals);
-            Runtime.XDecref(locals);
-            if (result == IntPtr.Zero)
+            bool borrowedLocals = true;
+            if (locals == null)
             {
-                return null;
+                locals = Runtime.PyDict_New();
+                borrowedLocals = false;
             }
-            return new PyObject(result);
-        }
 
-        public static PyObject RunString(string code, IntPtr globals, IntPtr locals)
-        {
             IntPtr flag = (IntPtr)257; /* Py_file_input */
-            IntPtr result = Runtime.PyRun_String(code, flag, globals, locals);
-            if (result == IntPtr.Zero)
+
+            try
             {
-                return null;
+                IntPtr result = Runtime.PyRun_String(
+                    code, flag, globals.Value, locals.Value
+                    );
+
+                if (Runtime.PyErr_Occurred() != 0)
+                {
+                    throw new PythonException();
+                }
+
+                return new PyObject(result);
```

Comment:
It was more of a caveat, return statements within `try` statements confuse me at times.